topics/RDMA.md

```diff
@@ -0,0 +1,170 @@
+Valkey module that can be dynamically loaded on demand.
+
+## Getting Started
+
+RDMA enables direct data exchange between networked computers' main memory,
+bypassing processors and operating systems.
+
+As a result, RDMA offers better performance compared to TCP. Test results indicate that
+Valkey Over RDMA achieves approximately 2 times higher QPS and lower latency.
+
+Please note that Valkey Over RDMA is currently supported only on Linux.
+
+## Running manually
+
+To run a Valkey server with RDMA mode:
+
+    ./src/valkey-server --protected-mode no \
+         --loadmodule src/valkey-rdma.so bind=192.168.122.100 port=6379
+
+Bind address/port of RDMA can be modified at runtime using the following command:
+
+    192.168.122.100:6379> CONFIG SET rdma-port 6380
+
+Valkey can run both RDMA and TCP concurrently on the same port:
+
+    ./src/valkey-server --protected-mode no \
+         --loadmodule src/valkey-rdma.so bind=192.168.122.100 port=6379 \
+         --port 6379
+
+Note that the network interface (192.168.122.100 of this example) should support
+RDMA. To test a server supports RDMA or not:
+
+    ~# rdma dev show (a new version iproute2 package)
+Or:
+
+    ~# ibv_devices (ibverbs-utils package of Debian/Ubuntu)
+
+
+## Protocol
+
+The protocol defines the QP type RC (like TCP), communication commands, and payload exchange mechanism.
+This dependency is based solely on the RDMA (aka Infiniband) specification
+and is independent of both software (including the OS and user libraries)
+and hardware (including vendors and low-level transports).
+
+Valkey Over RDMA has control-plane (control messages) and data-plane (payload transfer).
+
+### Control message
+
+Control messages use fixed 32-byte big-endian message structures:
+```C
+typedef struct ValkeyRdmaFeature {
+    /* defined as following Opcodes */
+    uint16_t opcode;
+    /* select features */
+    uint16_t select;
+    uint8_t rsvd[20];
+    /* feature bits */
+    uint64_t features;
+} ValkeyRdmaFeature;
+
+typedef struct ValkeyRdmaKeepalive {
+    /* defined as following Opcodes */
+    uint16_t opcode;
+    uint8_t rsvd[30];
+} ValkeyRdmaKeepalive;
+
+typedef struct ValkeyRdmaMemory {
+    /* defined as following Opcodes */
+    uint16_t opcode;
+    uint8_t rsvd[14];
+    /* address of a transfer buffer which is used to receive remote streaming data,
+     * aka 'RX buffer address'. The remote side should use this as 'TX buffer address' */
+    uint64_t addr;
+    /* length of the 'RX buffer' */
+    uint32_t length;
+    /* the RDMA remote key of 'RX buffer' */
+    uint32_t key;
+} ValkeyRdmaMemory;
+
+typedef union ValkeyRdmaCmd {
+    ValkeyRdmaFeature feature;
+    ValkeyRdmaKeepalive keepalive;
+    ValkeyRdmaMemory memory;
+} ValkeyRdmaCmd;
+```
+
+### Opcodes
+|Command| Value | Description |
+| :----: | :----: | :----: |
+| `GetServerFeature`   | 0 | required, get the features offered by Valkey server |
```

Comment:
What does required mean in this context? All of the opcodes are apparently required. There are also op codes mentioned later, are they same or different opcodes?

Reply:
Yes, the whole commands are required in the current document. As time goes on, more features may be introduced into the specification, then client&server use `GetServerFeature` and `SetClientFeature` to negotiate features. The negotiatable features are optional.

A typical usage in the future, VALKEY_RDMA_FEATURE_FOO and `ValkeyRdmaFoo` are defined:
* an old client connects a new server, detect `VALKEY_RDMA_FEATURE_FOO` but not use.
* a new client connects a new server, detect `VALKEY_RDMA_FEATURE_FOO` and use `ValkeyRdmaFoo` command
* a new client connects a old server, no `VALKEY_RDMA_FEATURE_FOO` detected, not use `ValkeyRdmaFoo`

Then we can have an extensible specification, and avoid unrecognized commands for both client and server.